synapse_antispam/mjolnir/antispam.py

```diff
@@ -14,115 +14,145 @@
+
+    def _update_rules(self):
+        async def run():
+            try:
+                logger.info("Updating mjolnir rules...")
+
+                # HACK: Private member access (_hs)
+                resp = await self._api._hs.get_proxied_http_client().get_json(self._config['rules_url'])
+
+                # *** !! DANGER !! ***
+                # We're about to trust that the admin has secured their network appropriately and
+                # that the values returned by the configurable URL are safe to execute as real
+                # python code. We're using eval(), which should mean that it can only be expressions,
+                # however there is still risk of damages. We are knowingly processing the strings
+                # returned and assuming they won't try to take control of the homeserver, or worse.
+                # *** !! DANGER !! ***
+
+                self._code_spam_checks = self._compile_rules(resp['checks']['spam'])
+                self._code_invite_checks = self._compile_rules(resp['checks']['invites'])
+                self._code_profile_checks = self._compile_rules(resp['checks']['profiles'])
+                self._code_create_room_checks = self._compile_rules(resp['checks']['createRoom'])
+                self._code_create_alias_checks = self._compile_rules(resp['checks']['createAlias'])
+                self._code_publish_room_checks = self._compile_rules(resp['checks']['publishRoom'])
+
+                logger.info("Updated mjolnir rules")
+            except Exception as e:
+                logger.warning("Error updating mjolnir rules: %s", e)
+
+        return defer.ensureDeferred(run())
+
+    def _compile_rules(self, rules):
+        return [{
+            "search": compile(rule["search"], '<string>', 'eval'),
+            "pattern": rule["pattern"],
+        } for rule in rules]
 
     # --- spam checker interface below here ---
 
     def check_event_for_spam(self, event):
-        room_id = event.get("room_id", "")
-        event_type = event.get("type", "")
-        state_key = event.get("state_key", None)
-
-        # Rebuild the rules if there's an event for our ban lists
-        if state_key is not None and event_type in ALL_RULE_TYPES and room_id in self.list_room_ids:
-            logger.info("Received ban list event - updating list")
-            self.get_list_for_room(room_id).build(with_event=event)
-            return False  # Ban list updates aren't spam
-
-        if not self.block_messages:
-            return False  # not spam (we aren't blocking messages)
-
-        sender = UserID.from_string(event.get("sender", ""))
-        if self.is_user_banned(sender.to_string()):
-            return True
-        if self.is_server_banned(sender.domain):
-            return True
+        for check in self._code_spam_checks:
+            params = {
+                "event": event,
+                "UserID": UserID,
+            }
+            search = eval(check["search"], {}, params)
```

Comment:
This can throw if the received code doesn't make sense with the input parameters. Right now this will raise an exception, is that the proper behavior?